internal/persiststorage/persiststorage_configmap.go

```diff
@@ -249,3 +253,45 @@ func (s *KubeConfigMapStore) ProcessChanges(ctx context.Context, dataMap **map[s
 
 	return
 }
+
+func (s *KubeConfigMapStore) ProcessChangesWithFunction(ctx context.Context, function ProcessFunc) (err error) {
+	raw_opt := metav1.SingleObject(metav1.ObjectMeta{
+		Namespace: s.namespace,
+		Name:      s.name,
+	})
+	opt := clnt.ListOptions{}
+	opt.Raw = &raw_opt
+
+	watcher, err := s.client.Watch(ctx, &corev1.ConfigMapList{}, &opt)
+
+	go func() {
+		for {
+			event, open := <-watcher.ResultChan()
+			if open {
+				switch event.Type {
+				case watch.Added, watch.Modified:
+					configmap, _ := event.Object.(*corev1.ConfigMap)
+					newMap := map[string]data.Object{}
+					for k, value := range configmap.Data {
+						var object data.Object
+						err = (*s.jsonAPI).Unmarshal([]byte(value), &object)
```

Comment:
Why the explicit pointer dereference? Should work without it: `s.jsonAPI.Unmarshal(...)`

Reply:
```internal/persiststorage/persiststorage_configmap.go:277:23: s.jsonAPI.Unmarshal undefined (type *jsoniter.API is pointer to interface, not interface)
make: *** [Makefile:292: binary] Error 1```
The persistent storage keep a pointer to the same jsonAPI object used by its owner handler

Reply:
Ah, I understand now. `s.jsonAPI` should not be a pointer because the target is an interface and in Go variables of interface types are already pointers. So you should to change the `jsonAPI` field of the `KubeConfigMapStore` struct from `*jsoniter.API` to just `jsoniter.API`. Then here you can use directly `s.jsonAPI.Unmarshal(...)`.

Reply:
I do have an strong opinion . Not that it introduces a bug, but it is really uncommon and unnecessary to use a pointer to an interface in general, and in this case in particular.

When you declare a variable like `var jsonAPI jsoniter.API` what the Go compiler will do is create a structure with two pointers: one pointer to the table describing the methods of the interface (know as an _itable_) and another pointer to the object itself. When you then do something like `jsonAPI = x` what you are doing is copying those two pointers, not creating a new object.

If you do `var jsonAPI *jsoniter.API` and then `jsonAPI = &x` what you are doing is copying the address of the `x` variable. That isn't necessary in general, and in particular not necessary in this case. I addition it means that the `x` object can't be stored in the stack because the Go compiler can't know if it will be used after the current scope is closed: that is known in Go as "escaping" to the heap. Finally it means that you can do `*jsonAPI = &y` and the result will be that the code that uses `x` would now surprisingly be using `y` instead.

Reply:
@jhernand Thanks for long explanation. Really appreciate it! Will change it later today.